Source/Core/Core/IOS/USB/Emulated/Skylander.cpp

```diff
@@ -0,0 +1,815 @@
+  m_endpoint_descriptor.emplace_back(EndpointDescriptor{0x7, 0x5, 0x81, 0x3, 0x40, 0x1});
+  m_endpoint_descriptor.emplace_back(EndpointDescriptor{0x7, 0x5, 0x2, 0x3, 0x40, 0x1});
+}
+
+SkylanderUSB::~SkylanderUSB()
+{
+}
+
+DeviceDescriptor SkylanderUSB::GetDeviceDescriptor() const
+{
+  return m_device_descriptor;
+}
+
+std::vector<ConfigDescriptor> SkylanderUSB::GetConfigurations() const
+{
+  return m_config_descriptor;
+}
+
+std::vector<InterfaceDescriptor> SkylanderUSB::GetInterfaces(u8 config) const
+{
+  return m_interface_descriptor;
+}
+
+std::vector<EndpointDescriptor> SkylanderUSB::GetEndpoints(u8 config, u8 interface, u8 alt) const
+{
+  return m_endpoint_descriptor;
+}
+
+bool SkylanderUSB::Attach()
+{
+  if (m_device_attached)
+  {
+    return true;
+  }
+  DEBUG_LOG_FMT(IOS_USB, "[{:04x}:{:04x}] Opening device", m_vid, m_pid);
+  m_device_attached = true;
+  return true;
+}
+
+bool SkylanderUSB::AttachAndChangeInterface(const u8 interface)
+{
+  return true;
+}
+
+int SkylanderUSB::CancelTransfer(const u8 endpoint)
+{
+  INFO_LOG_FMT(IOS_USB, "[{:04x}:{:04x} {}] Cancelling transfers (endpoint {:#x})", m_vid, m_pid,
+               m_active_interface, endpoint);
+
+  return IPC_SUCCESS;
+}
+
+int SkylanderUSB::ChangeInterface(const u8 interface)
+{
+  DEBUG_LOG_FMT(IOS_USB, "[{:04x}:{:04x} {}] Changing interface to {}", m_vid, m_pid,
+                m_active_interface, interface);
+  m_active_interface = interface;
+  return 0;
+}
+
+int SkylanderUSB::GetNumberOfAltSettings(u8 interface)
+{
+  return 0;
+}
+
+int SkylanderUSB::SetAltSetting(u8 alt_setting)
+{
+  return 0;
+}
+
+// Skylander Portal control transfers are handled via this method, if a transfer requires a
+// response, then we request that data from the "Portal" and queue it to be responded to via the
+// Interrupt Response, references can be found via:
+// https://marijnkneppers.dev/posts
+// https://github.com/tresni/PoweredPortals/wiki/USB-Protocols
+// https://pastebin.com/EqtTRzeF
+int SkylanderUSB::SubmitTransfer(std::unique_ptr<CtrlMessage> cmd)
+{
+  DEBUG_LOG_FMT(IOS_USB,
+                "[{:04x}:{:04x} {}] Control: bRequestType={:02x} bRequest={} wValue={:04x}"
+                " wIndex={:04x} wLength={:04x}",
+                m_vid, m_pid, m_active_interface, cmd->request_type, cmd->request, cmd->value,
+                cmd->index, cmd->length);
+
+  auto& system = Core::System::GetInstance();
+  auto& memory = system.GetMemory();
+  u8* buf = memory.GetPointerForRange(cmd->data_address, cmd->length);
+  if ((cmd->length == 0 || buf == nullptr) && cmd->request == 0x09)
+  {
+    ERROR_LOG_FMT(IOS_USB, "Skylander command invalid");
+    return IPC_EINVAL;
+  }
+  std::array<u8, 64> result = {};
+  std::array<u8, 64> data = {};
+  s32 expected_count = 0;
+  u64 expected_time = 100;
+  // Control transfers are instantaneous
+  switch (cmd->request_type)
+  {
+  // HID host to device type
+  case 0x21:
```

Comment:
Consider turning this outer level switch into an if. At one point you have three deeply nested switch statements, which makes this kinda hard to follow.